Phase-based operations

dasanchez opened the pull request:
- Added a Delegator class
- Added a Validator class
- delegator and validator replacements are now done with string replacements (via `sed`)
- Categorized operations into string-replacement (sed) and json ones
- User must submit a set of tasks that conform to the right order (string replacement first, followed by json operations)
- Added tests with pytest

dasanchez committed:
First commit

dasanchez committed:
Returns invalid sequence on bad user list

dasanchez committed:
Enable workflows

dasanchez committed:
Lint check main module only

dasanchez committed:
Python 3.10 for GH action

dasanchez committed:
Fix syntax

dasanchez committed:
Run dash

dasanchez committed:
Disable import error for pylint

dasanchez committed:
Add pytest to workflow

dasanchez committed:
Install dependencies

dasanchez committed:
Renamed workflow

dasanchez committed:
Run stateful genesis test

dasanchez committed:
Test fresh genesis only

dasanchez committed:
Add expected order to invalid sequence response

dasanchez committed:
Simplify workflow

dasanchez committed:
Update uses syntax

dasanchez committed:
uses updates

dasanchez committed:
Named steps

dasanchez committed:
Missing dashes

dasanchez committed:
Renamed workflow

uditvira commented on README.md:
```diff
@@ -27,38 +25,42 @@ Full documentation for the module can be accessed with `pydoc cosmos_genesis_tin
+    old_delegator=old_del,
+    new_delegator=new_del)
+    
+tinker.add_task(tinker.increase_balance,
```
love this api

uditvira commented on cosmos_genesis_tinker.py:
```diff
@@ -205,19 +488,43 @@ def generate_shasum(self):
         """
-        Save a modified genesis file back to JSON
+        Look through .app_state.auth.accounts,
+        check for "name": "bonded_tokens_pool"
```
nice! this is much better than hardcoding

uditvira commented on legacy/genesis_theta_local.py:
```diff
@@ -61,19 +61,25 @@
-subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_SELF_DELEGATION_ADDR    + "%" + LOCAL_SELF_DELEGATION_ADDR   + "%g",PREPROCESSED_FILE])
-subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_SELF_DELEGATION_PUBKEY  + "%" + LOCAL_SELF_DELEGATION_PUBKEY + "%g",PREPROCESSED_FILE])
-subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_VALIDATOR_OP_ADDRESS    + "%" + LOCAL_VALIDATOR_OP_ADDRESS   + "%g",PREPROCESSED_FILE])
+subprocess.call(["sed", "-i", ".bak", "s%" + COINBASE_SELF_DELEGATION_ADDR +
```
should we remove the legacy folder @dasanchez? not sure what the value in keeping it is?

dasanchez replied:
I'd moved the old examples to a legacy folder just to have them as reference, but yes, I'd rather delete them instead.

dasanchez replied:
Deleted the legacy folder and turned the theta tinker script into a stateful genesis example.

uditvira commented on cosmos_genesis_tinker.py:
```diff
@@ -1,108 +1,385 @@
+            print('Invalid sequence: replace_validator and replace_delegator '
+                  'must come before all other functions.')
+            print('Expected order:', str([task.func.__name__ for task in self._task_list.tasks()]))
+            return True
```
great!

uditvira reviewed (approved):
This is very impressive work. Thank you @dasanchez 

You probably already did, but curious if you explicitly compared shasum from legacy version and this new version?

dasanchez commented:
> curious if you explicitly compared shasum from legacy version and this new version?
@uditvira Good point, I did not check the shasum.
I will run the mainnet exported genesis through both versions and compare the shasum between the two before I merge.

dasanchez committed:
Re-ordered inserted dict in create_coin

dasanchez committed:
Use theta testnet genesis as example

dasanchez committed:
Delete legacy folder

dasanchez commented:
@uditvira 
The shasum didn't match when comparing a theta testnet genesis file generated with the old API to one generated with the new API.

The genesis file lists balances with amount first, and denom second: https://github.com/hyphacoop/cosmos-genesis-tinkerer/blob/312ef6dde0443663dde7962c3a97e033080f90cf/tests/fresh_genesis.json#L119

When I was writing the tests I noticed the `create_coin` function had those reversed (denom first, amount second): https://github.com/hyphacoop/cosmos-genesis-tinkerer/blob/59c4fc08a6bcad9600d286932ec985abd187d23b/cosmos_genesis_tinker.py#L433

So I changed this function at the time to match the genesis file order. That caused the shasum mismatch!

I have updated `create_coin` again to match the original version (denom first, amount second): https://github.com/hyphacoop/cosmos-genesis-tinkerer/blob/312ef6dde0443663dde7962c3a97e033080f90cf/cosmos_genesis_tinker.py#L627

I don't think gaiad will care if these are reversed when it's parsing the json.

I have removed the legacy folder and updated the theta genesis script to use the new API as an example for stateful genesis tinkering.

**The shasums now match**
